### PR TITLE
Add client_credentials grant type support

### PR DESCRIPTION
**Pull request by alek-sys**
Fixes gh-51

### CODE REVIEW

**Comment by alek-sys** [comment]
Hey @jgrandja, I still have a couple of to-dos:
- [ ] Support basic auth for client id / secret
- [ ] Update sample app to use JWK to sign the token. This should allow to put together an end-to-end demo

With that in mind, feel free to provide any initial feedback you have.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository;
+import org.springframework.security.oauth2.server.authorization.token.TokenIssuer;
+
+/**
+ * An {@link AuthenticationProvider} that converts give {@link OAuth2ClientCredentialsAuthenticationToken} to a {@link OAuth2AccessTokenAuthenticationToken}
+ * using the provided token generator.
+ *
+ * @author Alexey Nesterov
+ */
+public class OAuth2ClientCredentialsAuthenticationProvider implements AuthenticationProvider {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java:34]: #72 introduced `OAuth2ClientAuthenticationProvider` and `OAuth2ClientAuthenticationToken`, which handles client authentication. The `OAuth2ClientCredentialsAuthenticationProvider` should simply assert that the current `Authentication` is `OAuth2ClientAuthenticationToken`. Take a look at `OAuth2AuthorizationCodeAuthenticationProvider` and apply the same validation check for `OAuth2ClientAuthenticationToken`.

Also, take a look at the integration test `OAuth2AuthorizationCodeGrantTests.requestWhenTokenRequestValidThenResponseIncludesCacheHeaders()` to see how the client authenticates via Basic before it gets to `OAuth2AuthorizationCodeAuthenticationProvider`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/token/TokenIssuer.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization.token;
+
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+
+/**
+ * @author Alexey Nesterov
+ */
+public interface TokenIssuer<PRINCIPAL> {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/token/TokenIssuer.java:23]: Please remove this as it's not needed as part of this issue. The `OAuth2ClientCredentialsAuthenticationProvider` should simply leverage `StringKeyGenerator accessTokenGenerator` the same as in `OAuth2AuthorizationCodeAuthenticationProvider`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		if (!clientSecret.equals(registeredClient.getClientSecret())) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_CLIENT));
+		}
+
+		return new OAuth2AccessTokenAuthenticationToken(registeredClient, token, this.tokenIssuer.issue(registeredClient));
+	}
+
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return authentication == OAuth2ClientCredentialsAuthenticationToken.class;
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java:64]: Change to `return OAuth2ClientCredentialsAuthenticationToken.class.isAssignableFrom(authentication)`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java [hunk 4]
```diff
@@ -86,8 +84,10 @@
 	private final AuthenticationManager authenticationManager;
 	private final OAuth2AuthorizationService authorizationService;
 	private final RequestMatcher tokenEndpointMatcher;
-	private final Converter<HttpServletRequest, Authentication> authorizationGrantAuthenticationConverter =
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:89]: The way I envisioned `authorizationGrantAuthenticationConverter` is to be a delegation-based implementation that simply delegates to the specific authorization grant implementation. Spring Security uses this strategy in a number of places. Take a look at `DelegatingAccessDeniedHandler` and `DelegatingOAuth2UserService` for examples.

For now, both `AuthorizationCodeAuthenticationConverter` and `ClientCredentialsAuthenticationConverter` should be both `private static class` and then create a `delegate` `Converter` that simply delegates to the appropriate impl.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/AuthorizationCodeAuthenticationConverter.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import org.springframework.security.oauth2.core.AuthorizationGrantType;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.security.oauth2.core.endpoint.OAuth2ParameterNames;
+import org.springframework.security.oauth2.server.authorization.authentication.OAuth2AuthorizationCodeAuthenticationToken;
+import org.springframework.util.MultiValueMap;
+import org.springframework.util.StringUtils;
+
+class AuthorizationCodeAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/AuthorizationCodeAuthenticationConverter.java:33]: Please revert this change and keep `AuthorizationCodeAuthenticationConverter` as `private static class` in `OAuth2TokenEndpointFilter`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		if (!StringUtils.hasText(grantType) ||
+				parameters.get(OAuth2ParameterNames.GRANT_TYPE).size() != 1) {
+			throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.GRANT_TYPE);
+		}
+		if (!AuthorizationGrantType.CLIENT_CREDENTIALS.getValue().equals(grantType)) {
+			throwError(OAuth2ErrorCodes.UNSUPPORTED_GRANT_TYPE, OAuth2ParameterNames.GRANT_TYPE);
+		}
+
+		// client_id (REQUIRED)
+		String clientId = parameters.getFirst(OAuth2ParameterNames.CLIENT_ID);
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java:49]: `OAuth2ClientAuthenticationFilter` and `OAuth2ClientAuthenticationProvider` will handle authenticating the client. Both `client_id` and `client_secret` should be removed from the validation checks. Take a look at `AuthorizationCodeAuthenticationConverter` and how it returns `clientPrincipal = SecurityContextHolder.getContext().getAuthentication()` in the `Authentication` and apply the same.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+			throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.CLIENT_ID);
+		}
+
+		// client_secret (REQUIRED)
+		String clientSecret = parameters.getFirst(OAuth2ParameterNames.CLIENT_SECRET);
+		if (!StringUtils.hasText(clientSecret)) {
+			throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames.CLIENT_SECRET);
+		}
+
+		return new OAuth2ClientCredentialsAuthenticationToken(clientId, clientSecret);
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java:60]: `scope` is an optional parameter but should also be returned in `OAuth2ClientCredentialsAuthenticationToken`

**Comment by jgrandja** [comment]
@alek-sys 

Regarding

> Support basic auth for client id / secret

This is already implemented in `ClientSecretBasicAuthenticationConverter`

**Comment by alek-sys** [comment]
Thanks for feedback @jgrandja! I'm now re-working this PR based on your comments.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java (hunk 7 above)
Comment (alek-sys) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java]: Do you expect section [3.3. Access Token](https://tools.ietf.org/html/rfc6749#section-3.3) to be implemented as a part of this PR? E.g. returning `scope` in the response, and filtering requested scopes based on a list of scopes supported by a client?

Comment (jgrandja) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientCredentialsAuthenticationConverter.java]: Yes. The `OAuth2ClientCredentialsAuthenticationProvider` should receive the "request scopes" and the provider should ensure the "request scopes"  are either a subset (or all) of the "assigned scopes". If the client requests a scope that is not in the "assigned scopes" then it should fail. If no scopes are requested then we should auto-approve. However, I'm not sure auto-approving is a good idea but we can change that later if need be.

**Comment by alek-sys** [comment]
Updated based on your comments.

Also, [the spec](https://tools.ietf.org/html/rfc6749#section-3.3) states:
> If the issued access token scope is different from the one requested by the client, the authorization server MUST include the "scope" response parameter to inform the client of the actual scope granted.

But currently Token Endpoint always returns `scope` in the response. Shoud l I fix it in this PR or create a separate story for this (unless there is one already)?

**Comment by jgrandja** [comment]
The returned scope will come into play in #42 at a later point. No need to address it in this PR. But feel free to add notes to #42.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java [hunk 8]
```diff
@@ -214,4 +222,57 @@ public Authentication convert(HttpServletRequest request) {
...
+			// "scope" is optional, see https://tools.ietf.org/html/rfc6749#section-4.4.2 and https://tools.ietf.org/html/rfc6749#section-3.3
+			// if "scope" is not provided, client scopes are used
+			String scopeParameter = request.getParameter(SCOPE_PARAMETER_NAME);
+			if (StringUtils.isEmpty(scopeParameter)) {
+				return new OAuth2ClientCredentialsAuthenticationToken(clientAuthenticationToken, clientScopes);
+			}
+
+			Set<String> requestedScopes = null;
+			try {
+				requestedScopes = scopesParser.parse(scopeParameter);
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:265]: Please move `scope` validation in `OAuth2ClientCredentialsAuthenticationProvider`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+		Converter<HttpServletRequest, Authentication> converter = this.converters.get(new AuthorizationGrantType(grantType));
+		if (converter == null) {
+			throwError(OAuth2ErrorCodes.UNSUPPORTED_GRANT_TYPE, "grant_type");
+		}
+
+		return converter.convert(request);
+	}
+
+	public DelegateGrantTypeAuthenticationConverter registerConverter(AuthorizationGrantType authorizationGrantType, Converter<HttpServletRequest, Authentication> authenticationConverter) {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:49]: Please remove this method and  provide constructor that accepts the `Map` making it immutable.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.core.AuthorizationGrantType;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+import org.springframework.util.StringUtils;
+
+public class DelegateGrantTypeAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:30]: Rename to `DelegatingAuthorizationGrantAuthenticationConverter` and make `final`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+	private final HashMap<AuthorizationGrantType, Converter<HttpServletRequest, Authentication>> converters = new HashMap<>();
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:32]: Reference `Map` instead

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+	@Override
+	public Authentication convert(HttpServletRequest request) {
+		String grantType = request.getParameter("grant_type");
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:36]: This class should not throw any errors. It simply delegates to the correct `Converter` or returns `null` . Similar to how `DelegatingOAuth2UserService` is implemented.

Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:36]: Use constants in `OAuth2ParameterNames`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+ * An {@link AuthenticationProvider} that converts give {@link OAuth2ClientAuthenticationToken} to a {@link OAuth2AccessTokenAuthenticationToken}
+ * using the provided token generator.
+ *
+ * @author Alexey Nesterov
+ */
+public class OAuth2ClientCredentialsAuthenticationProvider implements AuthenticationProvider {
+
+	private final StringKeyGenerator accessTokenGenerator = new Base64StringKeyGenerator(Base64.getUrlEncoder());
+
+	public OAuth2ClientCredentialsAuthenticationProvider() {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java:42]: Please remove default constructor since it's redundant

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+		return "";
+	}
+
+	@Override
+	public Object getPrincipal() {
+		return this.clientPrincipal;
+	}
+
+	public Set<String> getScopes() {
+		return scopes;
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:55]: Prefix `this.`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java [hunk 15]
```diff
@@ -214,4 +222,57 @@ public Authentication convert(HttpServletRequest request) {
...
+				return null;
+			}
+
+			if (!OAuth2ClientAuthenticationToken.class.isAssignableFrom(authentication.getClass())) {
+				throwError(OAuth2ErrorCodes.INVALID_CLIENT, "client_id");
+			}
+
+			OAuth2ClientAuthenticationToken clientAuthenticationToken = (OAuth2ClientAuthenticationToken) authentication;
+			if (!clientAuthenticationToken.isAuthenticated()) {
+				throwError(OAuth2ErrorCodes.INVALID_CLIENT, "client_id");
```
[the remark this answers is not in the document]
Comment (alek-sys) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java]: I see now. Converter belongs to `web` package so it shouldn't really cotain any auth-n logic. So I refactored it a bit - now converted only extracts scopes from the request plus gets current authentication from the context. The rest is done in the manager, including validating scopes, using default scopes and validating client authentication. It is much more clear now, thanks for feedback.

**Comment by alek-sys** [comment]
Thanks @jgrandja, I've updated this PR. Now there is much better split between converted and provider.

**Comment by jgrandja** [comment]
Great thanks @alek-sys. I will review this tomorrow.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java [hunk 16]
```diff
@@ -214,4 +222,57 @@ public Authentication convert(HttpServletRequest request) {
...
+	/**
+	 * Validates that client is authenticated by {@link OAuth2ClientAuthenticationFilter}
+	 *
+	 * @see OAuth2ClientAuthenticationToken
+	 * @see OAuth2ClientAuthenticationFilter
+	 * @see <a target="_blank" href="https://tools.ietf.org/html/rfc6749#section-4.4">Section 4.4 Client Credentials Grant</a>
+	 */
+	private static class ClientCredentialsAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
+
+		private static final String SCOPE_PARAMETER_NAME = "scope";
```
[the remark this answers is not in the document]
Comment (jgrandja) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java]: Please remove `SCOPE_PARAMETER_NAME` and use `OAuth2ParameterNames.SCOPE` in `convert`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java (hunk 10 above)
Comment (jgrandja) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:29]: Looks like this was missed

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java (hunk 8 above)
Comment (jgrandja) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:251]: Actually the more I thought about this one, I don't think we need to validate here. The `scope` validation should be done when the `RegisteredClient` is created - similar to `ClientRegistration` in `oauth2-client`. We know then that the configured scopes are valid and all we need to do is match the configured ones against the requested ones (as-is). So let's remove `OAuth2ScopeParser` and pass in the requested scopes as is, eg:

`new HashSet<>(Arrays.asList(StringUtils.delimitedListToStringArray(scope, " ")))`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java (hunk 10 above)
Comment (alek-sys) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DelegateGrantTypeAuthenticationConverter.java:29]: Sorry for that, I missed this comment totally - for whatever reason it is not shown on `Conversations` tab, only on `Files`. So I get notifcations about your comments but have to search for them. New GitHub UI is not so polished I guess. This class is now updated.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java (hunk 8 above)
Comment (alek-sys) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:251]: The validation itself (e.g. check that client has all requested scopes) is happening in the provider. The purpose of`OAuth2ScopeParser`  is only parsing `scope` string request param to a set according to RFC. It is slightly more complicated than `delimitedListToStringArray` so I decided to introduce more strict parser to cover special characters, double spaces, etc. I think it still has some value, unless you absolutely want to remove it. 

Maybe exception for cases when `scope` is not a valid string should be different then, just general BAD_REQUEST?

**Comment by alek-sys** [comment]
@jgrandja updated based on your latest comments. Also, see my comment regarding `OAuth2ScopeParser` - I think it has some value as it validates `scope` parameter to be compliant with RFC.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jgrandja) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java]: Looks like this was missed

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java (hunk 8 above)
Comment (jgrandja) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java:256]: I still feel that `scope` validation should happen when a client is registered, eg. when a `RegisteredClient` is created. This ensures that the `scope`(s) the client is allowed to request `RegisteredClient.scopes` are valid from the start. Even if the requested `scope` has invalid characters, it still won't match against the registered (valid) scope. So I feel this extra validation at runtime is redundant so please remove `OAuth2ScopeParser`. FYI, this validation already exists in `RegisteredClient.validateScopes()`

**Comment by jgrandja** [comment]
@alek-sys Can you please update the test methods to following the naming convention `**methodName**When**Condition**Then**Result**`. Take a look at `OAuth2TokenEndpointFilterTests` and `OAuth2AuthorizationCodeAuthenticationProviderTests` for examples.

Also, please add integration tests in `OAuth2ClientCredentailsGrantTests` (new). See `OAuth2AuthorizationCodeGrantTests` for example.

**Comment by alek-sys** [comment]
@jgrandja done. Actually there was `OAuth2ClientCredentialsGrantTests` class with an integration test already. I updated it a bit to make it more like `OAuth2AuthorizationCodeGrantTests`. I also changed security configuration for it to be more realistic by overriding authentication entry point to return 401 when client is not authenticated and `client_credentials` grant is used.

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+package org.springframework.security.oauth2.server.authorization.authentication;
+
+import java.util.Collections;
+import java.util.Set;
+
+import org.springframework.security.authentication.AbstractAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.util.Assert;
+
+public class OAuth2ClientCredentialsAuthenticationToken extends AbstractAuthenticationToken {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:26]: - Add class-level javadoc with `@author` and `@since`
- Add `OAuth2ClientCredentialsAuthenticationTokenTests`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+	private final Authentication clientPrincipal;
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:28]: Add `private static final long serialVersionUID = Version.SERIAL_VERSION_UID;`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+	private final Set<String> scopes;
+
+	public OAuth2ClientCredentialsAuthenticationToken(OAuth2ClientAuthenticationToken clientPrincipal, Set<String> scopes) {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:31]: Use `Authentication` type instead of `OAuth2ClientAuthenticationToken `

Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:31]: Update to `@Nullable Set<String> scopes` and remove the other constructor

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+	public Object getCredentials() {
+		return "";
+	}
+
+	@Override
+	public Object getPrincipal() {
+		return this.clientPrincipal;
+	}
+
+	public Set<String> getScopes() {
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:54]: Add `@Nullable`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.OAuth2Error;
+import org.springframework.security.oauth2.core.OAuth2ErrorCodes;
+
+/**
+ * An {@link AuthenticationProvider} that converts give {@link OAuth2ClientAuthenticationToken} to a {@link OAuth2AccessTokenAuthenticationToken}
+ * using the provided token generator.
+ *
+ * @author Alexey Nesterov
```
Comment (jgrandja) [inline core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProvider.java:37]: Add `@since`

File: core/src/test/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProviderTest.java [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
+import org.springframework.security.oauth2.server.authorization.client.TestRegisteredClients;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class OAuth2ClientCredentialsAuthenticationProviderTest {
```
Comment (jgrandja) [inline core/src/test/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationProviderTest.java:32]: Rename to `...Tests`

File: core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.when;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.httpBasic;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+public class OAuth2ClientCredentialsGrantTests {
```
Comment (jgrandja) [inline core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java:48]: Add `@author`

File: core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	@Before
+	public void setup() {
+		reset(registeredClientRepository);
+		reset(authorizationService);
+	}
+
+	@Test
+	public void requestWhenAuthorizationRequestAuthenticatedThenThenReturnTokenAndScope() throws Exception {
```
Comment (jgrandja) [inline core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java:72]: Change "AuthorizationRequest" to "TokenRequest"

File: core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+					.with(csrf())
+					.param("grant_type", "client_credentials")
+					.param("scope", "email openid"))
+				.andExpect(status().isOk())
+				.andExpect(jsonPath("$.access_token").isNotEmpty())
+				.andExpect(jsonPath("$.scope").value("openid email"));
+	}
+
+	@Test
+	public void requestWhenAuthorizationRequestNotAuthenticatedThenUnauthorized() throws Exception {
```
Comment (jgrandja) [inline core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java:89]: Change "AuthorizationRequest" to "TokenRequest"

File: core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+				.thenReturn(client);
+
+		this.mvc.perform(post(OAuth2TokenEndpointFilter.DEFAULT_TOKEN_ENDPOINT_URI)
+				.param("grant_type", "client_credentials")
+				.param("scope", "email openid"))
+				.andExpect(status().isUnauthorized());
+	}
+
+	@EnableWebSecurity
+	static class AuthorizationServerConfiguration extends WebSecurityConfigurerAdapter {
```
Comment (jgrandja) [inline core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientCredentialsGrantTests.java:102]: Please change this to:

```java
@EnableWebSecurity
@Import(OAuth2AuthorizationServerConfiguration.class)
static class AuthorizationServerConfiguration {
...
```

See `OAuth2AuthorizationCodeGrantTests`

Any changes needed to the default security should be made in `OAuth2AuthorizationServerSecurity`

File: core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java (hunk 20 above)
Comment (alek-sys) [reply core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientCredentialsAuthenticationToken.java:65]: `scopes` is never null, do you mean `@NonNull`?

**Comment by alek-sys** [comment]
All updated based on your review @jgrandja. I only didn't add `@Nullable` since scope field is never null. Also, I noticed there is no logging anywhere in the project. Is that intentional or should I add at least debug logging to some of the components?

**Comment by jgrandja** [comment]
Thanks for adding this feature @alek-sys !

FYI, I followed up with a polish commit as some changes were needed and I wanted to get this merged.
Let me know if you have any questions regarding the changes in the polish commit.
